storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java

```diff
@@ -64,302 +64,395 @@
+            .stream()
+            .filter(e -> !assignedPartitions.contains(e))
+            .collect(Collectors.toSet());
+        unassignedPartitions.forEach(unassignedPartition -> {
+            remotePartitionMetadataEventHandler.clearTopicPartition(unassignedPartition);
+            readOffsetsByUserTopicPartition.remove(unassignedPartition);
+        });
+        log.info("Unassigned user-topic-partitions: {}", unassignedPartitions.size());
     }
 
-    public void addAssignmentsForPartitions(Set<TopicIdPartition> partitions) {
-        updateAssignmentsForPartitions(partitions, Collections.emptySet());
+    public void addAssignmentsForPartitions(final Set<TopicIdPartition> partitions) {
+        updateAssignments(Objects.requireNonNull(partitions), Collections.emptySet());
     }
 
-    public void removeAssignmentsForPartitions(Set<TopicIdPartition> partitions) {
-        updateAssignmentsForPartitions(Collections.emptySet(), partitions);
+    public void removeAssignmentsForPartitions(final Set<TopicIdPartition> partitions) {
+        updateAssignments(Collections.emptySet(), Objects.requireNonNull(partitions));
     }
 
-    private void updateAssignmentsForPartitions(Set<TopicIdPartition> addedPartitions,
-                                                Set<TopicIdPartition> removedPartitions) {
-        log.info("Updating assignments for addedPartitions: {} and removedPartition: {}", addedPartitions, removedPartitions);
+    private void updateAssignments(final Set<TopicIdPartition> addedPartitions,
+                                   final Set<TopicIdPartition> removedPartitions) {
+        log.info("Updating assignments for partitions added: {} and removed: {}", addedPartitions, removedPartitions);
+        if (!addedPartitions.isEmpty() || !removedPartitions.isEmpty()) {
+            synchronized (assignPartitionsLock) {
+                // Make a copy of the existing assignments and update the copy.
+                final Map<TopicIdPartition, UserTopicIdPartition> updatedUserPartitions = new HashMap<>(assignedUserTopicIdPartitions);
+                addedPartitions.forEach(tpId -> updatedUserPartitions.putIfAbsent(tpId, newUserTopicIdPartition(tpId)));
+                removedPartitions.forEach(updatedUserPartitions::remove);
+                if (!updatedUserPartitions.equals(assignedUserTopicIdPartitions)) {
+                    assignedUserTopicIdPartitions = Collections.unmodifiableMap(updatedUserPartitions);
+                    hasAssignmentChanged = true;
+                    log.debug("Assigned user-topic-partitions: {}", assignedUserTopicIdPartitions);
+                    assignPartitionsLock.notifyAll();
+                }
+            }
+        }
+    }
 
-        Objects.requireNonNull(addedPartitions, "addedPartitions must not be null");
-        Objects.requireNonNull(removedPartitions, "removedPartitions must not be null");
+    public Optional<Long> receivedOffsetForPartition(final int partition) {
```

Comment:
nit: reuse latest concept (and all usage methods called the same)
```suggestion
    public Optional<Long> latestOffsetForMetadataPartition(final int partition) {
```

Reply:
Changed to readXXX format as discussed in another comment